### ISSUE TITLE
defaults for deserialize?

### ISSUE THREAD

**Issue by agilgur5** [mobxjs/serializr#27]
From testing, it seems like if the given JSON passed as the second argument to `deserialize` has properties that are not explicitly listed on the schema, they aren't used/set at all.

In my case, I only list a few properties (dates, identifiers, references, and objects) and tend not to list every single primitive property that may appear.

I would like for there to be an option to make properties that aren't explicitly listed to just be treated as primitives.

For example:
```
class User {
  @serialize(identifier())
  id = null
  @serializable(date())
  join_date = null
  @serializable(list(object(Post)))
  favorites = null

  fullAddress () {
    return this.address + ', ' + this.city + ' ' + this.state
  }
}

let user = deserialize(User, {
  id: 2,
  join_date: '2012-04-12',
  favorites: [{link: 'foo.org', title: 'what a nice bar'}],
  settings: {can_post: true, gets_notifications: false},
  address: '1 Test St',
  city: 'Test',
  state: 'TT'
})
```
I'd like for `user` to have `settings`, `address`, etc, but right now the only way of doing this seems to be to explicitly list every single property

Is there a way to do this right now? Could it be supported somehow?

**Comment by mweststrate** [comment]
A default attribute serializer is supported by using the `*` matcher (see [readme](https://github.com/mobxjs/serializr#propschema)). However, there is no convenient way to express that you want to use that one with decorators. I think the following should do the job:

```
class User {
  // as is
}

getDefaultModelSchema(User).props["*"] = true
````

**Comment by agilgur5** [comment]
I see, I wasn't sure what the `*` matcher was for as it's under `PropSchema` but not for an individual field. Thanks for the clarification and example!

For any future readers, it could also be implemented as something like this:
```
function decorator (cls) {
  getDefaultModelSchema(class).props['*'] = true
  return cls
}
```
for reusability.

Not sure if it's really appropriate to implement inside the library.

**Comment by agilgur5** [comment]
@mweststrate so I ran into another problem that the `*` matcher doesn't seem to handle: lists & maps of primitives throw errors.

For instance. the `*` matcher won't match a propSchema like this: `@serializable(list(map(primitive())))`. An example of where this occurs would be if I removed the `Post` modelSchema reference in my initial example under `favorites` -- when `favorites` is parsed it would throw an error if there is no propSchema defined, even if the `*` matcher is in place.

The error is:
```Uncaught Error: [serializr] encountered non primitive value while deserializing '*' properties in property 'favorites':```

Explicitly stating `@serializable(list(map(primitive())))` also isn't possible in my scenario as `favorites` aren't always passed and may be `undefined`, so I'd have to create a custom propSchema to handle both cases (which so far feels like it may be a common enough occurrence to support an `or` or `if` propSchema helper for conditional propSchemas)

**Comment by mweststrate** [comment]
@agilgur5 feel free to create a PR with the additional decorator! Not sure about the name. Maybe something explicit like `@autoSerializePrimitives`. But it's longish...

The invariant is probably too strict, as also mentioned in #34. I think either the `*` serializer should not even try to serialize non-primitive values, or serialize them as is and assert they contain valid json structures.

**Comment by mweststrate** [comment]
Introduced `@serializeAll` in 1.1.11. Also the `*` decorator will now skip non-primitives during serialization

**Comment by agilgur5** [comment]
`@serializeAll` seems like a misnomer -- I think I preferred `@autoSerializePrimitives` for its explicitness. I'm leaning towards asserting JSON structures (both for deserialize and serialize), though I think skipping is still better than throwing.

I mainly use this for deserialization -- so in my case I solved my problem of `@serializable(list(map(primitive()))) favorites` throwing by simply adding a default `= []` to the end (that was a little unintuitive to realize).

And oh no, I missed a chance to contribute :(. Moved to GitLab at work and missed some notifications during that migration
